File: Day09/Python/que3linkedlist-reverse.py
class SinglyList:
    class Node:
        def __init__(self, val):
            self.data = val
            self.next = None

    def __init__(self):
        self.head = None

    def display(self):
        print("List: ", end="")
        trav = self.head
        while trav is not None:
            print(trav.data, "-> ", end="")
            trav = trav.next
        print()

    def addLast(self, val):
        newNode = self.Node(val)
        if self.head is None:
            self.head = newNode
        else:
            trav = self.head
            while trav.next is not None:
                trav = trav.next
            trav.next = newNode

    def revList(self):
        oldhead = self.head
        self.head = None
        while oldhead is not None:
            temp = oldhead
            oldhead = oldhead.next
            temp.next = self.head
            self.head = temp

    def recRevList(self, trav=None):
        if trav is None:
            if self.head is not None:
                self.recRevList(self.head)
            return
        if trav.next is None:
            self.head = trav
            return trav
        self.recRevList(trav.next).next = trav
        trav.next = None
        return trav


def main():
    list = SinglyList()
    list.addLast(10)
    list.addLast(20)
    list.addLast(30)
    list.addLast(40)
    list.addLast(50)
    list.display()
    # list.revList()    # list reverse without recursion
    list.recRevList()  # list reverse using recursion
    list.display()


if __name__ == "__main__":
    main()

File: Day09/C++/que3linkedlist_reverse.cpp
#include <iostream>
using namespace std;

class SinglyList
{
private:
    class Node
    {
    public:
        int data;
        Node *next;
        Node(int val) : data(val), next(nullptr) {}
    };

    Node *head;

public:
    SinglyList() : head(nullptr) {}

    void display()
    {
        cout << "List: ";
        Node *trav = head;
        while (trav != nullptr)
        {
            cout << trav->data << " -> ";
            trav = trav->next;
        }
        cout << endl;
    }

    void addLast(int val)
    {
        Node *newNode = new Node(val);
        if (head == nullptr)
        {
            head = newNode;
        }
        else
        {
            Node *trav = head;
            while (trav->next != nullptr)
            {
                trav = trav->next;
            }
            trav->next = newNode;
        }
    }

    void revList()
    {
        Node *oldhead = head;
        head = nullptr;
        while (oldhead != nullptr)
        {
            Node *temp = oldhead;
            oldhead = oldhead->next;
            temp->next = head;
            head = temp;
        }
    }

    Node *recRevList(Node *trav)
    {
        if (trav->next == nullptr)
        {
            head = trav;
            return trav;
        }
        recRevList(trav->next)->next = trav;
        trav->next = nullptr;
        return trav;
    }

    void recRevList()
    {
        if (head != nullptr)
        {
            recRevList(head);
        }
    }

    ~SinglyList()
    {
        Node *trav = head;
        while (trav != nullptr)
        {
            Node *temp = trav;
            trav = trav->next;
            delete temp;
        }
    }
};

int main()
{
    SinglyList list;
    list.addLast(10);
    list.addLast(20);
    list.addLast(30);
    list.addLast(40);
    list.addLast(50);
    list.display();
    // list.revList();    // list reverse without recursion
    list.recRevList(); // list reverse using recursion
    list.display();
    return 0;
}